File: buildy_lib/toolchain.py
"""Toolchain management and command building"""

import logging
import yaml
from pathlib import Path
from dataclasses import dataclass
from typing import Dict, List, Optional, Any

logger = logging.getLogger('buildy.toolchain')

@dataclass
class Tool:
    """Individual tool within a toolchain"""
    name: str
    action: str
    command: str
    input_extensions: List[str]
    output_extension: str
    output_pattern: str = "{name}"
    flags: Dict[str, List[str]] = None
    supports: Dict[str, Any] = None
    
    def __post_init__(self):
        if self.flags is None:
            self.flags = {}
        if self.supports is None:
            self.supports = {}

@dataclass
class ToolchainConfig:
    """Toolchain configuration loaded from YAML"""
    name: str
    description: str
    target_platform: str
    target_architecture: str
    host_platform: str
    host_architecture: str
    execution_type: str
    execution_config: Dict[str, Any]
    tools: Dict[str, Tool]  # tool_name -> Tool object
    
    @staticmethod
    def load(toolchain_file: str) -> 'ToolchainConfig':
        """Load toolchain configuration from YAML"""
        try:
            with open(toolchain_file, 'r') as f:
                data = yaml.safe_load(f)
            
            tc = data['toolchain']
            
            # Parse tools
            tools = {}
            for tool_name, tool_data in tc.get('tools', {}).items():
                # Normalize input_extensions to list
                input_exts = tool_data.get('input_extensions', [])
                if isinstance(input_exts, str):
                    input_exts = [input_exts]
                
                tools[tool_name] = Tool(
                    name=tool_name,
                    action=tool_data.get('action', 'compile'),
                    command=tool_data.get('command', ''),
                    input_extensions=input_exts,
                    output_extension=tool_data.get('output_extension', ''),
                    output_pattern=tool_data.get('output_pattern', '{name}'),
                    flags=tool_data.get('flags', {}),
                    supports=tool_data.get('supports', {})
                )
            
            return ToolchainConfig(
                name=tc['name'],
                description=tc.get('description', ''),
                target_platform=tc['target']['platform'],
                target_architecture=tc['target']['architecture'],
                host_platform=tc['host']['platform'],
                host_architecture=tc['host']['architecture'],
                execution_type=tc['execution']['type'],
                execution_config=tc['execution'],
                tools=tools
            )
        except Exception as e:
            logger.error(f"Failed to load toolchain from {toolchain_file}: {e}")
            raise

class ToolMatcher:
    """Matches files to appropriate tools based on action and extension"""
    
    def __init__(self, toolchain: ToolchainConfig):
        self.toolchain = toolchain
        # Build lookup table: (action, extension) -> Tool
        self._tool_map: Dict[tuple[str, str], Tool] = {}
        self._build_tool_map()
    
    def _build_tool_map(self):
        """Build action+extension lookup map"""
        for tool in self.toolchain.tools.values():
            # Only build map for non-link actions
            # Link tools are handled separately by find_link_tool()
            if tool.action != 'link':
                for ext in tool.input_extensions:
                    key = (tool.action, ext)
                    if key in self._tool_map:
                        existing = self._tool_map[key]
                        logger.error(
                            f"Ambiguous tool definition in toolchain '{self.toolchain.name}': "
                            f"Both '{existing.name}' and '{tool.name}' handle action='{tool.action}' "
                            f"with extension='{ext}'"
                        )
                        raise ValueError(f"Ambiguous tool mapping: {key}")
                    self._tool_map[key] = tool
    
    def find_tool(self, action: str, file_path: str) -> Optional[Tool]:
        """Find tool that matches action and file extension
        
        Args:
            action: The action to perform (e.g., 'compile', 'link', 'convert')
            file_path: Path to the file (extension will be extracted)
        
        Returns:
            Matching Tool or None if no match found
        """
        ext = Path(file_path).suffix
        if not ext:
            return None
        
        key = (action, ext)
        tool = self._tool_map.get(key)
        
        if tool:
            logger.debug(f"Matched {file_path} ({action}) -> tool '{tool.name}'")
        else:
            logger.debug(f"No tool found for action='{action}' extension='{ext}'")
        
        return tool
    
    def find_link_tool(self, output_type: str) -> Optional[Tool]:
        """Find tool for linking based on output type
        
        Args:
            output_type: 'shared_library', 'static_library', or 'executable'
        
        Returns:
            Matching Tool or None
        """
        # Look for tool with action='link' and matching output type
        for tool in self.toolchain.tools.values():
            if tool.action == 'link':
                # Check if tool name or supports indicates it handles this output type
                if output_type in tool.name or tool.supports.get('output_type') == output_type:
                    logger.debug(f"Matched link tool for '{output_type}' -> '{tool.name}'")
                    return tool
        
        logger.warning(f"No link tool found for output_type='{output_type}'")
        return None

class ToolchainManager:
    """Manages toolchain selection and loading"""
    
    def __init__(self, toolchains_dir: str = "toolchains"):
        self.toolchains_dir = Path(toolchains_dir)
        self._toolchains: Dict[str, ToolchainConfig] = {}
        self._load_toolchains()
    
    def _load_toolchains(self):
        """Load all toolchain configurations"""
        if not self.toolchains_dir.exists():
            logger.warning(f"Toolchains directory not found: {self.toolchains_dir}")
            return
        
        for tc_file in self.toolchains_dir.glob("*.yaml"):
            try:
                tc = ToolchainConfig.load(tc_file)
                self._toolchains[tc.name] = tc
                logger.debug(f"Loaded toolchain: {tc.name} - {tc.description}")
            except Exception as e:
                logger.error(f"Failed to load toolchain {tc_file}: {e}")
    
    def get_toolchain(self, name: str) -> Optional[ToolchainConfig]:
        """Get toolchain by name"""
        return self._toolchains.get(name)
    
    def auto_detect(self, platform: str, architecture: str) -> Optional[ToolchainConfig]:
        """Auto-detect best toolchain for platform/architecture"""
        # Try to find matching native toolchain
        for tc in self._toolchains.values():
            if (tc.target_platform == platform and 
                tc.target_architecture == architecture and
                tc.execution_type == 'native'):
                logger.info(f"Auto-detected toolchain: {tc.name}")
                return tc
        
        logger.warning(f"No native toolchain found for {platform}-{architecture}")
        return None
    
    def list_toolchains(self) -> List[tuple[str, str]]:
        """List available toolchains with descriptions"""
        return [(tc.name, tc.description) for tc in self._toolchains.values()]

class CommandBuilder:
    """Builds commands using tool-based approach"""
    
    def __init__(self, toolchain: ToolchainConfig, tool_matcher: ToolMatcher, config_type: str = 'debug'):
        self.toolchain = toolchain
        self.tool_matcher = tool_matcher
        self.config_type = config_type
    
    def build_command(self, tool: Tool, source: str, output: str,
                     defines: List[str] = None, include_dirs: List[str] = None,
                     extra_flags: List[str] = None, **kwargs) -> tuple[str, str]:
        """Build command using tool template
        
        Args:
            tool: The Tool to use
            source: Input file path
            output: Output file path
            defines: Preprocessor defines (if tool supports)
            include_dirs: Include directories (if tool supports)
            extra_flags: Additional flags
            **kwargs: Additional template variables (std, pic, etc.)
        
        Returns:
            tuple: (command, dep_file) - The command and dependency file path (if applicable)
        """
        defines = defines or []
        include_dirs = include_dirs or []
        extra_flags = extra_flags or []
        
        # Normalize extra_flags to list if it's a string
        if isinstance(extra_flags, str):
            extra_flags = [extra_flags] if extra_flags else []
        
        # Get flags for current configuration
        common_flags = tool.flags.get('common', [])
        config_flags = tool.flags.get(self.config_type, [])
        all_flags = common_flags + config_flags + extra_flags
        
        # Build template variables
        template_vars = {
            'input': source,
            'output': output,
            'flags': ' '.join(all_flags),
        }
        
        # Add defines if tool supports them
        if tool.supports.get('defines') and defines:
            define_flag = tool.supports.get('define_flag', '-D')
            template_vars['defines'] = ' '.join(f"{define_flag}{d}" for d in defines)
        else:
            template_vars['defines'] = ''
        
        # Add includes if tool supports them
        if tool.supports.get('includes') and include_dirs:
            include_flag = tool.supports.get('include_flag', '-I')
            template_vars['includes'] = ' '.join(f"{include_flag}{inc}" for inc in include_dirs)
        else:
            template_vars['includes'] = ''
        
        # Add PIC flag if tool supports it and requested
        if tool.supports.get('pic') and kwargs.get('is_shared_library'):
            template_vars['pic'] = tool.supports.get('pic_flag', '-fPIC')
        else:
            template_vars['pic'] = ''
        
        # Handle dependency generation if tool supports it
        dep_file = ''
        if tool.supports.get('dependencies'):
            dep_template = tool.supports.get('dependencies')
            if dep_template:
                dep_file = output.replace(tool.output_extension, '.d')
                template_vars['dep_file'] = dep_file
                template_vars['dep_flags'] = dep_template.format(dep_file=dep_file)
            else:
                template_vars['dep_flags'] = ''
        else:
            template_vars['dep_flags'] = ''
        
        # Add any additional kwargs
        for key, value in kwargs.items():
            if key not in template_vars:
                template_vars[key] = value
        
        # Build command from template
        try:
            command = tool.command.format(**template_vars)
        except KeyError as e:
            logger.error(f"Missing template variable {e} for tool '{tool.name}'")
            raise
        
        # Clean up extra spaces
        command = ' '.join(command.split())
        
        return command, dep_file
    
    def build_link_command(self, tool: Tool, objects: List[str], output: str,
                          lib_dirs: List[str] = None, libs: List[str] = None) -> str:
        """Build link command using tool template
        
        Args:
            tool: The link tool to use
            objects: List of object files to link
            output: Output file path
            lib_dirs: Library search directories
            libs: Library names to link against
        """
        lib_dirs = lib_dirs or []
        libs = libs or []
        
        # Build template variables
        template_vars = {
            'objects': ' '.join(objects),
            'output': output,
        }
        
        # Add library directories if tool supports them
        if tool.supports.get('lib_dirs') and lib_dirs:
            lib_dir_flag = tool.supports.get('lib_dir_flag', '-L')
            template_vars['lib_dirs'] = ' '.join(f"{lib_dir_flag}{d}" for d in lib_dirs)
        else:
            template_vars['lib_dirs'] = ''
        
        # Add libraries if tool supports them
        if tool.supports.get('libs') and libs:
            lib_flag = tool.supports.get('lib_flag', '-l')
            if lib_flag:
                template_vars['libs'] = ' '.join(f"{lib_flag}{lib}" for lib in libs)
            else:
                # No prefix (e.g., MSVC style)
                template_vars['libs'] = ' '.join(libs)
        else:
            template_vars['libs'] = ''
        
        # Get flags for current configuration
        common_flags = tool.flags.get('common', [])
        config_flags = tool.flags.get(self.config_type, [])
        all_flags = common_flags + config_flags
        template_vars['flags'] = ' '.join(all_flags)
        
        # Build command
        try:
            command = tool.command.format(**template_vars)
        except KeyError as e:
            logger.error(f"Missing template variable {e} for link tool '{tool.name}'")
            raise
        
        # Clean up extra spaces
        command = ' '.join(command.split())
        
        return command

File: goBuildy/toolchain.go
package main

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

// Tool represents an individual tool within a toolchain
type Tool struct {
	Name            string              `yaml:"name"`
	Action          string              `yaml:"action"`
	Command         string              `yaml:"command"`
	InputExtensions []string            `yaml:"input_extensions"`
	OutputExtension string              `yaml:"output_extension"`
	OutputPattern   string              `yaml:"output_pattern"`
	Flags           map[string][]string `yaml:"flags"`
	Supports        map[string]any      `yaml:"supports"`
}

// NewTool creates a new Tool with defaults
func NewTool(name, action, command string, inputExts []string, outputExt string) *Tool {
	return &Tool{
		Name:            name,
		Action:          action,
		Command:         command,
		InputExtensions: inputExts,
		OutputExtension: outputExt,
		OutputPattern:   "{name}",
		Flags:           make(map[string][]string),
		Supports:        make(map[string]any),
	}
}

// ToolchainConfig represents a toolchain configuration loaded from YAML
type ToolchainConfig struct {
	Name               string           `yaml:"name"`
	Description        string           `yaml:"description"`
	TargetPlatform     string           `yaml:"-"`
	TargetArchitecture string           `yaml:"-"`
	HostPlatform       string           `yaml:"-"`
	HostArchitecture   string           `yaml:"-"`
	ExecutionType      string           `yaml:"-"`
	ExecutionConfig    map[string]any   `yaml:"-"`
	Tools              map[string]*Tool `yaml:"-"`
}

// NewToolchainConfig creates a new ToolchainConfig
func NewToolchainConfig(name, description string) *ToolchainConfig {
	return &ToolchainConfig{
		Name:            name,
		Description:     description,
		ExecutionConfig: make(map[string]any),
		Tools:           make(map[string]*Tool),
	}
}

// LoadToolchainConfig loads a toolchain configuration from a YAML file
func LoadToolchainConfig(toolchainFile string) (*ToolchainConfig, error) {
	data, err := os.ReadFile(toolchainFile)
	if err != nil {
		return nil, fmt.Errorf("failed to read toolchain file %s: %w", toolchainFile, err)
	}

	var rawConfig map[string]any
	if err := yaml.Unmarshal(data, &rawConfig); err != nil {
		return nil, fmt.Errorf("failed to parse toolchain YAML %s: %w", toolchainFile, err)
	}

	// Extract toolchain section
	tcData, ok := rawConfig["toolchain"].(map[string]any)
	if !ok {
		return nil, fmt.Errorf("toolchain file %s missing 'toolchain' section", toolchainFile)
	}

	tc := NewToolchainConfig("", "")

	// Parse basic fields
	if name, ok := tcData["name"].(string); ok {
		tc.Name = name
	} else {
		return nil, fmt.Errorf("toolchain missing 'name' field")
	}

	if desc, ok := tcData["description"].(string); ok {
		tc.Description = desc
	}

	// Parse target
	if targetData, ok := tcData["target"].(map[string]any); ok {
		if platform, ok := targetData["platform"].(string); ok {
			tc.TargetPlatform = platform
		}
		if arch, ok := targetData["architecture"].(string); ok {
			tc.TargetArchitecture = arch
		}
	}

	// Parse host
	if hostData, ok := tcData["host"].(map[string]any); ok {
		if platform, ok := hostData["platform"].(string); ok {
			tc.HostPlatform = platform
		}
		if arch, ok := hostData["architecture"].(string); ok {
			tc.HostArchitecture = arch
		}
	}

	// Parse execution
	if execData, ok := tcData["execution"].(map[string]any); ok {
		tc.ExecutionConfig = execData
		if execType, ok := execData["type"].(string); ok {
			tc.ExecutionType = execType
		}
	}

	// Parse tools
	if toolsData, ok := tcData["tools"].(map[string]any); ok {
		for toolName, toolDataRaw := range toolsData {
			toolData, ok := toolDataRaw.(map[string]any)
			if !ok {
				continue
			}

			tool := &Tool{
				Name:            toolName,
				Action:          "compile",
				OutputPattern:   "{name}",
				Flags:           make(map[string][]string),
				Supports:        make(map[string]any),
				InputExtensions: []string{},
			}

			// Parse tool fields
			if action, ok := toolData["action"].(string); ok {
				tool.Action = action
			}
			if command, ok := toolData["command"].(string); ok {
				tool.Command = command
			}
			if outputExt, ok := toolData["output_extension"].(string); ok {
				tool.OutputExtension = outputExt
			}
			if outputPattern, ok := toolData["output_pattern"].(string); ok {
				tool.OutputPattern = outputPattern
			}

			// Parse input_extensions (can be string or list)
			if inputExts, ok := toolData["input_extensions"]; ok {
				switch v := inputExts.(type) {
				case string:
					tool.InputExtensions = []string{v}
				case []any:
					for _, ext := range v {
						if extStr, ok := ext.(string); ok {
							tool.InputExtensions = append(tool.InputExtensions, extStr)
						}
					}
				}
			}

			// Parse flags
			if flagsData, ok := toolData["flags"].(map[string]any); ok {
				for flagType, flagsRaw := range flagsData {
					if flagsList, ok := flagsRaw.([]any); ok {
						flags := []string{}
						for _, flag := range flagsList {
							if flagStr, ok := flag.(string); ok {
								flags = append(flags, flagStr)
							}
						}
						tool.Flags[flagType] = flags
					}
				}
			}

			// Parse supports
			if supportsData, ok := toolData["supports"].(map[string]any); ok {
				tool.Supports = supportsData
			}

			tc.Tools[toolName] = tool
		}
	}

	return tc, nil
}

// LoadToolchainConfigFromDict creates a ToolchainConfig from a dictionary (for testing)
func LoadToolchainConfigFromDict(configDict map[string]any) (*ToolchainConfig, error) {
	tc := NewToolchainConfig("", "")

	if name, ok := configDict["name"].(string); ok {
		tc.Name = name
	} else {
		return nil, fmt.Errorf("toolchain missing 'name' field")
	}

	if desc, ok := configDict["description"].(string); ok {
		tc.Description = desc
	}

	if platform, ok := configDict["target_platform"].(string); ok {
		tc.TargetPlatform = platform
	} else {
		tc.TargetPlatform = "linux"
	}

	if arch, ok := configDict["target_architecture"].(string); ok {
		tc.TargetArchitecture = arch
	} else {
		tc.TargetArchitecture = "x86_64"
	}

	if platform, ok := configDict["host_platform"].(string); ok {
		tc.HostPlatform = platform
	} else {
		tc.HostPlatform = "linux"
	}

	if arch, ok := configDict["host_architecture"].(string); ok {
		tc.HostArchitecture = arch
	} else {
		tc.HostArchitecture = "x86_64"
	}

	if execType, ok := configDict["execution_type"].(string); ok {
		tc.ExecutionType = execType
	} else {
		tc.ExecutionType = "native"
	}

	tc.ExecutionConfig = map[string]any{"type": tc.ExecutionType}

	// Parse tools from list format
	if toolsList, ok := configDict["tools"].([]any); ok {
		for _, toolDataRaw := range toolsList {
			toolData, ok := toolDataRaw.(map[string]any)
			if !ok {
				continue
			}

			toolName, ok := toolData["name"].(string)
			if !ok {
				continue
			}

			tool := &Tool{
				Name:            toolName,
				Action:          "compile",
				OutputPattern:   "{name}",
				Flags:           make(map[string][]string),
				Supports:        make(map[string]any),
				InputExtensions: []string{},
			}

			if action, ok := toolData["action"].(string); ok {
				tool.Action = action
			}
			if command, ok := toolData["command"].(string); ok {
				tool.Command = command
			}
			if outputExt, ok := toolData["output_extension"].(string); ok {
				tool.OutputExtension = outputExt
			}
			if outputPattern, ok := toolData["output_pattern"].(string); ok {
				tool.OutputPattern = outputPattern
			}

			// Parse input_extensions
			if inputExts, ok := toolData["input_extensions"]; ok {
				switch v := inputExts.(type) {
				case string:
					tool.InputExtensions = []string{v}
				case []any:
					for _, ext := range v {
						if extStr, ok := ext.(string); ok {
							tool.InputExtensions = append(tool.InputExtensions, extStr)
						}
					}
				}
			}

			// Parse flags
			if flagsData, ok := toolData["flags"].(map[string]any); ok {
				for flagType, flagsRaw := range flagsData {
					if flagsList, ok := flagsRaw.([]any); ok {
						flags := []string{}
						for _, flag := range flagsList {
							if flagStr, ok := flag.(string); ok {
								flags = append(flags, flagStr)
							}
						}
						tool.Flags[flagType] = flags
					}
				}
			}

			// Parse supports
			if supportsData, ok := toolData["supports"].(map[string]any); ok {
				tool.Supports = supportsData
			}

			// Handle output_type for link tools
			if outputType, ok := toolData["output_type"].(string); ok {
				tool.Supports["output_type"] = outputType
			}

			tc.Tools[toolName] = tool
		}
	}

	return tc, nil
}

// ToolMatcher matches files to appropriate tools based on action and extension
type ToolMatcher struct {
	toolchain *ToolchainConfig
	toolMap   map[string]*Tool // key: "action:extension"
}

// NewToolMatcher creates a new ToolMatcher
func NewToolMatcher(toolchain *ToolchainConfig) (*ToolMatcher, error) {
	tm := &ToolMatcher{
		toolchain: toolchain,
		toolMap:   make(map[string]*Tool),
	}

	if err := tm.buildToolMap(); err != nil {
		return nil, err
	}

	return tm, nil
}

// buildToolMap builds the action+extension lookup map
func (tm *ToolMatcher) buildToolMap() error {
	for _, tool := range tm.toolchain.Tools {
		// Only build map for non-link actions
		// Link tools are handled separately by FindLinkTool()
		if tool.Action != "link" {
			for _, ext := range tool.InputExtensions {
				key := tool.Action + ":" + ext
				if existing, exists := tm.toolMap[key]; exists {
					return fmt.Errorf(
						"ambiguous tool definition in toolchain '%s': "+
							"both '%s' and '%s' handle action='%s' with extension='%s'",
						tm.toolchain.Name, existing.Name, tool.Name, tool.Action, ext,
					)
				}
				tm.toolMap[key] = tool
			}
		}
	}
	return nil
}

// FindTool finds a tool that matches the action and file extension
func (tm *ToolMatcher) FindTool(action, filePath string) *Tool {
	ext := filepath.Ext(filePath)
	if ext == "" {
		return nil
	}

	key := action + ":" + ext
	tool := tm.toolMap[key]

	if tool != nil {
		log.Printf("Matched %s (%s) -> tool '%s'", filePath, action, tool.Name)
	} else {
		log.Printf("No tool found for action='%s' extension='%s'", action, ext)
	}

	return tool
}

// FindLinkTool finds a tool for linking based on output type
func (tm *ToolMatcher) FindLinkTool(outputType string) *Tool {
	// Look for tool with action='link' and matching output type
	for _, tool := range tm.toolchain.Tools {
		if tool.Action == "link" {
			// Check if tool name or supports indicates it handles this output type
			if strings.Contains(tool.Name, outputType) {
				log.Printf("Matched link tool for '%s' -> '%s'", outputType, tool.Name)
				return tool
			}
			if supportedType, ok := tool.Supports["output_type"].(string); ok && supportedType == outputType {
				log.Printf("Matched link tool for '%s' -> '%s'", outputType, tool.Name)
				return tool
			}
		}
	}

	log.Printf("WARNING: No link tool found for output_type='%s'", outputType)
	return nil
}

// ToolchainManager manages toolchain selection and loading
type ToolchainManager struct {
	toolchainsDirs []string
	toolchains     map[string]*ToolchainConfig
}

// NewToolchainManager creates a new ToolchainManager from a single directory
func NewToolchainManager(toolchainsDir string) (*ToolchainManager, error) {
	return NewToolchainManagerMulti([]string{toolchainsDir})
}

// NewToolchainManagerMulti creates a new ToolchainManager from multiple directories
func NewToolchainManagerMulti(toolchainsDirs []string) (*ToolchainManager, error) {
	tm := &ToolchainManager{
		toolchainsDirs: toolchainsDirs,
		toolchains:     make(map[string]*ToolchainConfig),
	}

	if err := tm.loadToolchains(); err != nil {
		return nil, err
	}

	return tm, nil
}

// loadToolchains loads all toolchain configurations from all toolchain directories
func (tm *ToolchainManager) loadToolchains() error {
	// Load toolchains from each directory in order
	for _, toolchainsDir := range tm.toolchainsDirs {
		// Check if directory exists
		if _, err := os.Stat(toolchainsDir); os.IsNotExist(err) {
			log.Printf("WARNING: Toolchains directory not found: %s", toolchainsDir)
			continue // Not fatal, try next directory
		}

		// Read all .yaml files
		entries, err := os.ReadDir(toolchainsDir)
		if err != nil {
			log.Printf("WARNING: Failed to read toolchains directory %s: %v", toolchainsDir, err)
			continue
		}

		for _, entry := range entries {
			if entry.IsDir() {
				continue
			}

			if filepath.Ext(entry.Name()) == ".yaml" {
				tcFile := filepath.Join(toolchainsDir, entry.Name())
				tc, err := LoadToolchainConfig(tcFile)
				if err != nil {
					log.Printf("ERROR: Failed to load toolchain %s: %v", tcFile, err)
					continue
				}

				if _, exists := tm.toolchains[tc.Name]; exists {
					log.Printf("WARNING: Toolchain '%s' from %s overrides existing toolchain", tc.Name, toolchainsDir)
				}

				tm.toolchains[tc.Name] = tc
				log.Printf("Loaded toolchain: %s - %s", tc.Name, tc.Description)
			}
		}
	}

	return nil
}

// GetToolchain gets a toolchain by name
func (tm *ToolchainManager) GetToolchain(name string) *ToolchainConfig {
	return tm.toolchains[name]
}

// AutoDetect auto-detects the best toolchain for the given platform/architecture
func (tm *ToolchainManager) AutoDetect(platform, architecture string) *ToolchainConfig {
	// Collect matching toolchains
	var matches []*ToolchainConfig
	for _, tc := range tm.toolchains {
		if tc.TargetPlatform == platform &&
			tc.TargetArchitecture == architecture &&
			tc.ExecutionType == "native" {
			matches = append(matches, tc)
		}
	}

	if len(matches) == 0 {
		log.Printf("WARNING: No native toolchain found for %s-%s", platform, architecture)
		return nil
	}

	// Define platform-specific preferred toolchains
	preferredToolchains := map[string][]string{
		"linux":   {"gcc-linux", "clang-linux"},
		"windows": {"msvc-windows", "gcc-mingw", "clang-windows"},
		"macos":   {"clang-macos", "gcc-macos"},
		"darwin":  {"clang-macos", "gcc-macos"}, // darwin is macOS
	}

	// Try to find preferred toolchain for this platform
	if preferences, ok := preferredToolchains[platform]; ok {
		for _, preferredName := range preferences {
			for _, tc := range matches {
				if tc.Name == preferredName {
					log.Printf("Auto-detected toolchain: %s", tc.Name)
					return tc
				}
			}
		}
	}

	// Fallback: sort matches by name for deterministic selection
	sort.Slice(matches, func(i, j int) bool {
		return matches[i].Name < matches[j].Name
	})

	log.Printf("Auto-detected toolchain: %s (fallback)", matches[0].Name)
	return matches[0]
}

// ListToolchains lists all available toolchains with descriptions
func (tm *ToolchainManager) ListToolchains() []struct {
	Name        string
	Description string
} {
	result := []struct {
		Name        string
		Description string
	}{}

	for _, tc := range tm.toolchains {
		result = append(result, struct {
			Name        string
			Description string
		}{
			Name:        tc.Name,
			Description: tc.Description,
		})
	}

	return result
}

// CommandBuilder builds commands using tool-based approach
type CommandBuilder struct {
	toolchain   *ToolchainConfig
	toolMatcher *ToolMatcher
	configType  string
}

// NewCommandBuilder creates a new CommandBuilder
func NewCommandBuilder(toolchain *ToolchainConfig, toolMatcher *ToolMatcher, configType string) *CommandBuilder {
	return &CommandBuilder{
		toolchain:   toolchain,
		toolMatcher: toolMatcher,
		configType:  configType,
	}
}

// BuildCommand builds a command using the tool template
func (cb *CommandBuilder) BuildCommand(
	tool *Tool,
	source, output string,
	defines, includeDirs, extraFlags []string,
	kwargs map[string]any,
) (string, string, error) {
	if defines == nil {
		defines = []string{}
	}
	if includeDirs == nil {
		includeDirs = []string{}
	}
	if extraFlags == nil {
		extraFlags = []string{}
	}
	if kwargs == nil {
		kwargs = make(map[string]any)
	}

	// Get flags for current configuration
	commonFlags := tool.Flags["common"]
	configFlags := tool.Flags[cb.configType]
	allFlags := append([]string{}, commonFlags...)
	allFlags = append(allFlags, configFlags...)
	allFlags = append(allFlags, extraFlags...)

	// Build template variables
	templateVars := map[string]string{
		"source": source,
		"input":  source,
		"output": output,
		"flags":  strings.Join(allFlags, " "),
	}

	// Add defines if tool supports them
	if supportsDefines, ok := tool.Supports["defines"].(bool); ok && supportsDefines && len(defines) > 0 {
		defineFlag := "-D"
		if df, ok := tool.Supports["define_flag"].(string); ok {
			defineFlag = df
		}
		defineStrs := []string{}
		for _, d := range defines {
			defineStrs = append(defineStrs, defineFlag+d)
		}
		templateVars["defines"] = strings.Join(defineStrs, " ")
	} else {
		templateVars["defines"] = ""
	}

	// Add includes if tool supports them
	if supportsIncludes, ok := tool.Supports["includes"].(bool); ok && supportsIncludes && len(includeDirs) > 0 {
		includeFlag := "-I"
		if inf, ok := tool.Supports["include_flag"].(string); ok {
			includeFlag = inf
		}
		includeStrs := []string{}
		for _, inc := range includeDirs {
			includeStrs = append(includeStrs, includeFlag+inc)
		}
		templateVars["includes"] = strings.Join(includeStrs, " ")
	} else {
		templateVars["includes"] = ""
	}

	// Add PIC flag if tool supports it and requested
	if supportsPIC, ok := tool.Supports["pic"].(bool); ok && supportsPIC {
		if isShared, ok := kwargs["is_shared_library"].(bool); ok && isShared {
			picFlag := "-fPIC"
			if pf, ok := tool.Supports["pic_flag"].(string); ok {
				picFlag = pf
			}
			templateVars["pic"] = picFlag
		} else {
			templateVars["pic"] = ""
		}
	} else {
		templateVars["pic"] = ""
	}

	// Handle dependency generation if tool supports it
	depFile := ""
	if depTemplate, ok := tool.Supports["dependencies"].(string); ok && depTemplate != "" {
		// Use .json extension for MSVC /sourceDependencies, .d for GCC/Clang
		if strings.Contains(depTemplate, "/sourceDependencies") ||
			strings.Contains(strings.ToLower(depTemplate), "/sourcedependencies") {
			depFile = strings.TrimSuffix(output, tool.OutputExtension) + ".json"
		} else {
			depFile = strings.TrimSuffix(output, tool.OutputExtension) + ".d"
		}

		templateVars["dep_file"] = depFile
		templateVars["dep_flags"] = strings.ReplaceAll(depTemplate, "{dep_file}", depFile)
	} else {
		templateVars["dep_flags"] = ""
	}

	// Add any additional kwargs
	for key, value := range kwargs {
		if _, exists := templateVars[key]; !exists {
			templateVars[key] = fmt.Sprintf("%v", value)
		}
	}

	// Build command from template
	command := tool.Command
	for key, value := range templateVars {
		placeholder := "{" + key + "}"
		command = strings.ReplaceAll(command, placeholder, value)
	}

	command = strings.TrimSpace(command)

	return command, depFile, nil
}

// BuildLinkCommand builds a link command using the tool template
func (cb *CommandBuilder) BuildLinkCommand(
	tool *Tool,
	objects []string,
	output string,
	libDirs, libs, frameworks []string,
) (string, error) {
	if libDirs == nil {
		libDirs = []string{}
	}
	if libs == nil {
		libs = []string{}
	}
	if frameworks == nil {
		frameworks = []string{}
	}

	// Build template variables
	templateVars := map[string]string{
		"objects": strings.Join(objects, " "),
		"output":  output,
	}

	// Add library directories if tool supports them
	if supportsLibDirs, ok := tool.Supports["lib_dirs"].(bool); ok && supportsLibDirs && len(libDirs) > 0 {
		libDirFlag := "-L"
		if ldf, ok := tool.Supports["lib_dir_flag"].(string); ok {
			libDirFlag = ldf
		}
		libDirStrs := []string{}
		for _, d := range libDirs {
			libDirStrs = append(libDirStrs, libDirFlag+d)
		}
		templateVars["lib_dirs"] = strings.Join(libDirStrs, " ")
	} else {
		templateVars["lib_dirs"] = ""
	}

	// Add libraries if tool supports them
	if supportsLibs, ok := tool.Supports["libs"].(bool); ok && supportsLibs && len(libs) > 0 {
		libFlag := "-l"
		if lf, ok := tool.Supports["lib_flag"].(string); ok {
			libFlag = lf
		}

		if libFlag != "" {
			libStrs := []string{}
			for _, lib := range libs {
				libStrs = append(libStrs, libFlag+lib)
			}
			templateVars["libs"] = strings.Join(libStrs, " ")
		} else {
			// No prefix (e.g., MSVC style)
			templateVars["libs"] = strings.Join(libs, " ")
		}
	} else {
		templateVars["libs"] = ""
	}

	// Add frameworks if tool supports them (macOS only)
	if supportsFrameworks, ok := tool.Supports["frameworks"].(bool); ok && supportsFrameworks && len(frameworks) > 0 {
		frameworkFlag := "-framework "
		if ff, ok := tool.Supports["framework_flag"].(string); ok {
			frameworkFlag = ff
		}
		frameworkStrs := []string{}
		for _, fw := range frameworks {
			frameworkStrs = append(frameworkStrs, frameworkFlag+fw)
		}
		templateVars["frameworks"] = strings.Join(frameworkStrs, " ")
	} else {
		templateVars["frameworks"] = ""
	}

	// Get flags for current configuration
	commonFlags := tool.Flags["common"]
	configFlags := tool.Flags[cb.configType]
	allFlags := append([]string{}, commonFlags...)
	allFlags = append(allFlags, configFlags...)
	templateVars["flags"] = strings.Join(allFlags, " ")

	// Build command
	command := tool.Command
	for key, value := range templateVars {
		placeholder := "{" + key + "}"
		command = strings.ReplaceAll(command, placeholder, value)
	}

	command = strings.TrimSpace(command)

	return command, nil
}
